HBASE-28432 Refactor tools which are under test packaging to a new mo…

NihalJain opened the pull request:
…dule hbase-diagnostics

The purpose of this task is to refactor and move certain tools currently located under the test packaging to a new module, named 'hbase-diagnostics'.

The following tools have been initially identified for relocation(will add more as and when identified):
- [PerformanceEvaluation](https://github.com/apache/hbase/blob/master/hbase-mapreduce/src/test/java/org/apache/hadoop/hbase/PerformanceEvaluation.java)
- [LoadTestTool](https://github.com/apache/hbase/blob/936d267d1094e37222b9b836ab068689ccce3574/hbase-mapreduce/src/test/java/org/apache/hadoop/hbase/util/LoadTestTool.java)
- [HFilePerformanceEvaluation](https://github.com/apache/hbase/blob/936d267d1094e37222b9b836ab068689ccce3574/hbase-server/src/test/java/org/apache/hadoop/hbase/HFilePerformanceEvaluation.java)
- [ScanPerformanceEvaluation](https://github.com/apache/hbase/blob/936d267d1094e37222b9b836ab068689ccce3574/hbase-mapreduce/src/test/java/org/apache/hadoop/hbase/ScanPerformanceEvaluation.java)
- [LoadBalancerPerformanceEvaluation](https://github.com/apache/hbase/blob/936d267d1094e37222b9b836ab068689ccce3574/hbase-balancer/src/test/java/org/apache/hadoop/hbase/master/balancer/LoadBalancerPerformanceEvaluation.java)

These tools are valuable beyond the scope of testing and should be accessible in the binary distribution of HBase. However, their current location within the test jars adds unnecessary bloat to the assembly and classpath, and potentially introduces CVE-prone JARs into the binary assemblies. We plan to remove all test jars from assembly with [HBASE-28433](https://issues.apache.org/jira/browse/HBASE-28433).

This task involves creating the new 'hbase-diagnostics' module, and moving the identified tools into this module. It also includes ensuring that these tools function correctly in their new location and that their relocation does not negatively impact any existing functionality or dependencies.

Also see draft patch without this change for follow up work HBASE-28433 at https://github.com/apache/hbase/pull/6184

NihalJain commented:
very rough change. lot of things to handle:
- code duplication
- cleanup 
- organising methods into proper files
- testing
- also seems i didnot pull code before working here, so need to resolve those as well

Will post a summary of changes and things to be take care of later tomorrow

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 41s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m  3s |  |  master passed  |
| +1 :green_heart: |  compile  |   2m  6s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   4m  8s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 13s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   1m 32s | [/patch-mvninstall-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/patch-mvninstall-root.txt) |  root in the patch failed.  |
| -1 :x: |  compile  |   1m 22s | [/patch-compile-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/patch-compile-root.txt) |  root in the patch failed.  |
| -0 :warning: |  javac  |   1m 22s | [/patch-compile-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/patch-compile-root.txt) |  root in the patch failed.  |
| -0 :warning: |  javadoc  |   0m 13s | [/results-javadoc-javadoc-hbase-diagnostics.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/results-javadoc-javadoc-hbase-diagnostics.txt) |  hbase-diagnostics generated 3 new + 0 unchanged - 0 fixed = 3 total (was 0)  |
| -0 :warning: |  javadoc  |   2m  6s | [/results-javadoc-javadoc-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/results-javadoc-javadoc-root.txt) |  root generated 3 new + 91 unchanged - 0 fixed = 94 total (was 91)  |
| -1 :x: |  shadedjars  |   3m 51s |  |  patch has 16 errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  16m  9s | [/patch-unit-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-root.txt) |  root in the patch failed.  |
|  |   |  45m 44s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6249 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 1e96014cbaf4 5.4.0-195-generic #215-Ubuntu SMP Fri Aug 2 18:28:05 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / da6b8396d5e83f6bc7acf3e81a8d2ef5a8dbf3ce |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| shadedjars | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-jdk17-hadoop3-check/output/patch-shadedjars.txt |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/testReport/ |
| Max. process+thread count | 1495 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-client hbase-balancer hbase-asyncfs hbase-server hbase-mapreduce hbase-diagnostics . hbase-assembly hbase-it U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m 36s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m  3s |  |  master passed  |
| +1 :green_heart: |  compile  |   8m 10s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 13s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |  12m 13s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 44s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   1m 33s | [/patch-mvninstall-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-mvninstall-root.txt) |  root in the patch failed.  |
| -1 :x: |  compile  |   4m 17s | [/patch-compile-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-compile-root.txt) |  root in the patch failed.  |
| -0 :warning: |  javac  |   4m 17s | [/patch-compile-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-compile-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m  7s | [/results-checkstyle-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/results-checkstyle-root.txt) |  root: The patch generated 45 new + 49 unchanged - 36 fixed = 94 total (was 85)  |
| +1 :green_heart: |  xmllint  |   0m  0s |  |  No new issues.  |
| -1 :x: |  spotbugs  |   0m 34s | [/patch-spotbugs-hbase-server.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-spotbugs-hbase-server.txt) |  hbase-server in the patch failed.  |
| -1 :x: |  spotbugs  |   4m 12s | [/patch-spotbugs-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-spotbugs-root.txt) |  root in the patch failed.  |
| -1 :x: |  spotbugs  |   0m  7s | [/patch-spotbugs-hbase-it.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-spotbugs-hbase-it.txt) |  hbase-it in the patch failed.  |
| -1 :x: |  hadoopcheck  |   1m 50s |  |  The patch causes 16 errors with Hadoop v3.3.6.  |
| -1 :x: |  hadoopcheck  |   3m 41s |  |  The patch causes 16 errors with Hadoop v3.4.0.  |
| -1 :x: |  spotless  |   0m 37s |  |  patch has 24 errors when running spotless:check, run spotless:apply to fix.  |
|||| _ Other Tests _ |
| -1 :x: |  asflicense  |   1m 26s | [/results-asflicense.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/results-asflicense.txt) |  The patch generated 2 ASF License warnings.  |
|  |   |  52m  3s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6249 |
| Optional Tests | dupname asflicense javac codespell detsecrets xmllint spotless spotbugs checkstyle compile hadoopcheck hbaseanti |
| uname | Linux 9208861a2112 5.4.0-192-generic #212-Ubuntu SMP Fri Jul 5 09:47:39 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / da6b8396d5e83f6bc7acf3e81a8d2ef5a8dbf3ce |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| hadoopcheck | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-javac-3.3.6.txt |
| hadoopcheck | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-javac-3.4.0.txt |
| spotless | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/artifact/yetus-general-check/output/patch-spotless.txt |
| Max. process+thread count | 192 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-client hbase-balancer hbase-asyncfs hbase-server hbase-mapreduce hbase-diagnostics . hbase-assembly hbase-it U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/1/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 xmllint=20913 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

NihalJain commented on hbase-common/src/test/java/org/apache/hadoop/hbase/io/compress/CompressionTestBase.java:
```diff
@@ -26,7 +26,7 @@
 import org.apache.hadoop.hbase.util.Bytes;
 import org.apache.hadoop.hbase.util.EnvironmentEdgeManager;
-import org.apache.hadoop.hbase.util.RandomDistribution;
+import org.apache.hadoop.hbase.util.RandomDistributionCopy;
```
Created a copy just to bring out the all the usages of the class: RandomDistribution to help decide where we move this in final patch.

NihalJain commented:
Change Summary
- Added a new module called hbase-diagnostics (as suggested by Istvan)
  - Question: Any better names?
- Created a copy just to bring out the all the usages of the class to help decide where we move this in final patch:
  - RandomDistribution
  - KeyProviderForTesting
  - LoadTestKVGenerator
- Added a new util DiagnosticToolsCommonUtils.java and moved method required to loginAndReturnUGI from HBaseKerberosUtils.java
- Added a new util LoadTestUtil and moved all load test related code from HFileTestUtil and HBaseTestingUtil to here 
  - Question: Should we merge this to DiagnosticToolsCommonUtils?
- Copied generateData() from PerformanceEvaluation.java in TestHFileOutputFormat2.java to break cyclic dependency
- Added @InterfaceAudience.Private to all classes moved from test code to main
- Replaced UTIL.getConfiguration() with HBaseConfiguration.create() in LoadBalancerPerformanceEvaluation.java

TODO 
- Need to handle testHBASE14489() in hbase-server tests which depends on FilterAllFilter(), should we move this class to main of hbase-server or hbase-common?
- TODO: How to deal with copied classes which have cyclic dependencies, should copy to main of hbase-server or hbase-common?
- Some manual Testing
- Anything else which we need to handle?

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 38s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   4m 43s |  |  master passed  |
| +1 :green_heart: |  compile  |  12m 38s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 41s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |  18m 13s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 52s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 32s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 15s |  |  the patch passed  |
| -0 :warning: |  javac  |   8m 15s | [/results-compile-javac-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-general-check/output/results-compile-javac-root.txt) |  root generated 61 new + 1207 unchanged - 4 fixed = 1268 total (was 1211)  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m 12s | [/results-checkstyle-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-general-check/output/results-checkstyle-root.txt) |  root: The patch generated 43 new + 58 unchanged - 36 fixed = 101 total (was 94)  |
| +1 :green_heart: |  xmllint  |   0m  2s |  |  No new issues.  |
| -1 :x: |  spotbugs  |   7m 44s | [/new-spotbugs-root.html](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-general-check/output/new-spotbugs-root.html) |  root generated 11 new + 0 unchanged - 0 fixed = 11 total (was 0)  |
| +1 :green_heart: |  hadoopcheck  |  11m 47s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  spotless  |   0m 45s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   1m 33s |  |  The patch does not generate ASF License warnings.  |
|  |   |  90m  9s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:root |
|  |  Integral division result cast to double or float in org.apache.hadoop.hbase.PerformanceEvaluation.calculateRowsAndSize(PerformanceEvaluation$TestOptions)  At PerformanceEvaluation.java:double or float in org.apache.hadoop.hbase.PerformanceEvaluation.calculateRowsAndSize(PerformanceEvaluation$TestOptions)  At PerformanceEvaluation.java:[line 3154] |
|  |  org.apache.hadoop.hbase.PerformanceEvaluation$RunResult defines compareTo(PerformanceEvaluation$RunResult) and uses Object.equals()  At PerformanceEvaluation.java:Object.equals()  At PerformanceEvaluation.java:[line 250] |
|  |  Random object created and used only once in org.apache.hadoop.hbase.util.LoadTestKVGenerator.getValueForRowColumn(int, byte[][])  At LoadTestKVGenerator.java:only once in org.apache.hadoop.hbase.util.LoadTestKVGenerator.getValueForRowColumn(int, byte[][])  At LoadTestKVGenerator.java:[line 111] |
|  |  org.apache.hadoop.hbase.util.LoadTestTool.DEFAULT_NUM_REGIONS_PER_SERVER isn't final but should be  At LoadTestTool.java:be  At LoadTestTool.java:[line 163] |
|  |  org.apache.hadoop.hbase.util.LoadTestUtil.DEFAULT_COLUMN_FAMILY should be both final and package protected  At LoadTestUtil.java:and package protected  At LoadTestUtil.java:[line 48] |
|  |  org.apache.hadoop.hbase.util.MultiThreadedAction.verifyResultAgainstDataGenerator(Result, boolean, boolean) concatenates strings using + in a loop  At MultiThreadedAction.java:using + in a loop  At MultiThreadedAction.java:[line 414] |
|  |  Integral division result cast to double or float in org.apache.hadoop.hbase.util.MultiThreadedAction$ProgressReporter.run()  At MultiThreadedAction.java:double or float in org.apache.hadoop.hbase.util.MultiThreadedAction$ProgressReporter.run()  At MultiThreadedAction.java:[line 203] |
|  |  org.apache.hadoop.hbase.util.MultiThreadedReader$HBaseReaderThread.createGet(long) concatenates strings using + in a loop  At MultiThreadedReader.java:in a loop  At MultiThreadedReader.java:[line 318] |
|  |  Dead store to rowKey in org.apache.hadoop.hbase.util.MultiThreadedReaderWithACL$HBaseReaderThreadWithACL.queryKey(Get, boolean, long)  At MultiThreadedReaderWithACL.java:org.apache.hadoop.hbase.util.MultiThreadedReaderWithACL$HBaseReaderThreadWithACL.queryKey(Get, boolean, long)  At MultiThreadedReaderWithACL.java:[line 91] |
|  |  Inconsistent synchronization of org.apache.hadoop.hbase.util.MultiThreadedUpdater.writer; locked 75% of time  Unsynchronized access at MultiThreadedUpdater.java:75% of time  Unsynchronized access at MultiThreadedUpdater.java:[line 80] |
|  |  Unwritten field:MultiThreadedUpdaterWithACL.java:[line 94] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6249 |
| Optional Tests | dupname asflicense javac codespell detsecrets xmllint spotless spotbugs checkstyle compile hadoopcheck hbaseanti |
| uname | Linux 80ed0d4779d0 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 469be3d3d2e106f41c28d10938b296e9fb673ec6 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 190 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-client hbase-balancer hbase-asyncfs hbase-server hbase-mapreduce hbase-diagnostics hbase-compression/hbase-compression-zstd . hbase-assembly hbase-it U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 xmllint=20913 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 29s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m  1s |  |  master passed  |
| +1 :green_heart: |  compile  |   2m  5s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   3m 54s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 36s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 55s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m  8s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m  8s |  |  the patch passed  |
| -0 :warning: |  javadoc  |   0m 12s | [/results-javadoc-javadoc-hbase-diagnostics.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-jdk17-hadoop3-check/output/results-javadoc-javadoc-hbase-diagnostics.txt) |  hbase-diagnostics generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| -0 :warning: |  javadoc  |   1m 57s | [/results-javadoc-javadoc-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-jdk17-hadoop3-check/output/results-javadoc-javadoc-root.txt) |  root generated 2 new + 91 unchanged - 0 fixed = 93 total (was 91)  |
| +1 :green_heart: |  shadedjars  |   5m 33s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 245m 15s | [/patch-unit-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-root.txt) |  root in the patch failed.  |
|  |   | 282m 15s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6249 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 633e00b5ce0a 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 469be3d3d2e106f41c28d10938b296e9fb673ec6 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/testReport/ |
| Max. process+thread count | 6078 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-client hbase-balancer hbase-asyncfs hbase-server hbase-mapreduce hbase-diagnostics hbase-compression/hbase-compression-zstd . hbase-assembly hbase-it U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/2/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   2m 25s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m  2s |  |  master passed  |
| +1 :green_heart: |  compile  |   7m 56s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 13s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |  12m  0s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 44s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m  3s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m  1s |  |  the patch passed  |
| -0 :warning: |  javac  |   8m  1s | [/results-compile-javac-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-general-check/output/results-compile-javac-root.txt) |  root generated 61 new + 1207 unchanged - 4 fixed = 1268 total (was 1211)  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m 13s | [/results-checkstyle-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-general-check/output/results-checkstyle-root.txt) |  root: The patch generated 43 new + 58 unchanged - 36 fixed = 101 total (was 94)  |
| +1 :green_heart: |  xmllint  |   0m  0s |  |  No new issues.  |
| -1 :x: |  spotbugs  |   7m 42s | [/new-spotbugs-root.html](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-general-check/output/new-spotbugs-root.html) |  root generated 11 new + 0 unchanged - 0 fixed = 11 total (was 0)  |
| +1 :green_heart: |  hadoopcheck  |  11m 33s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  spotless  |   0m 43s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   1m 34s |  |  The patch does not generate ASF License warnings.  |
|  |   |  77m 17s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:root |
|  |  Integral division result cast to double or float in org.apache.hadoop.hbase.PerformanceEvaluation.calculateRowsAndSize(PerformanceEvaluation$TestOptions)  At PerformanceEvaluation.java:double or float in org.apache.hadoop.hbase.PerformanceEvaluation.calculateRowsAndSize(PerformanceEvaluation$TestOptions)  At PerformanceEvaluation.java:[line 3154] |
|  |  org.apache.hadoop.hbase.PerformanceEvaluation$RunResult defines compareTo(PerformanceEvaluation$RunResult) and uses Object.equals()  At PerformanceEvaluation.java:Object.equals()  At PerformanceEvaluation.java:[line 250] |
|  |  Random object created and used only once in org.apache.hadoop.hbase.util.LoadTestKVGenerator.getValueForRowColumn(int, byte[][])  At LoadTestKVGenerator.java:only once in org.apache.hadoop.hbase.util.LoadTestKVGenerator.getValueForRowColumn(int, byte[][])  At LoadTestKVGenerator.java:[line 111] |
|  |  org.apache.hadoop.hbase.util.LoadTestTool.DEFAULT_NUM_REGIONS_PER_SERVER isn't final but should be  At LoadTestTool.java:be  At LoadTestTool.java:[line 163] |
|  |  org.apache.hadoop.hbase.util.LoadTestUtil.DEFAULT_COLUMN_FAMILY should be both final and package protected  At LoadTestUtil.java:and package protected  At LoadTestUtil.java:[line 50] |
|  |  org.apache.hadoop.hbase.util.MultiThreadedAction.verifyResultAgainstDataGenerator(Result, boolean, boolean) concatenates strings using + in a loop  At MultiThreadedAction.java:using + in a loop  At MultiThreadedAction.java:[line 415] |
|  |  Integral division result cast to double or float in org.apache.hadoop.hbase.util.MultiThreadedAction$ProgressReporter.run()  At MultiThreadedAction.java:double or float in org.apache.hadoop.hbase.util.MultiThreadedAction$ProgressReporter.run()  At MultiThreadedAction.java:[line 206] |
|  |  org.apache.hadoop.hbase.util.MultiThreadedReader$HBaseReaderThread.createGet(long) concatenates strings using + in a loop  At MultiThreadedReader.java:in a loop  At MultiThreadedReader.java:[line 318] |
|  |  Dead store to rowKey in org.apache.hadoop.hbase.util.MultiThreadedReaderWithACL$HBaseReaderThreadWithACL.queryKey(Get, boolean, long)  At MultiThreadedReaderWithACL.java:org.apache.hadoop.hbase.util.MultiThreadedReaderWithACL$HBaseReaderThreadWithACL.queryKey(Get, boolean, long)  At MultiThreadedReaderWithACL.java:[line 91] |
|  |  Inconsistent synchronization of org.apache.hadoop.hbase.util.MultiThreadedUpdater.writer; locked 75% of time  Unsynchronized access at MultiThreadedUpdater.java:75% of time  Unsynchronized access at MultiThreadedUpdater.java:[line 80] |
|  |  Unwritten field:MultiThreadedUpdaterWithACL.java:[line 94] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6249 |
| Optional Tests | dupname asflicense javac codespell detsecrets xmllint spotless spotbugs checkstyle compile hadoopcheck hbaseanti |
| uname | Linux 162b34e96fd4 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 167c153ab498904c9361cc7d3b4a408ce45267b7 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 192 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-client hbase-balancer hbase-asyncfs hbase-server hbase-mapreduce hbase-diagnostics hbase-it hbase-compression/hbase-compression-zstd hbase-assembly . U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 xmllint=20913 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

stoty commented on hbase-common/src/test/java/org/apache/hadoop/hbase/util/RandomDistributionCopy.java:
```diff
@@ -0,0 +1,227 @@
+ * Remove after tfile is committed and use the tfile version of this class instead.
+ * </p>
+ */
+public class RandomDistributionCopy {
```
This is a bad name.

Do we need two copies of this class ? Can't we just move this from hbase-compression to hbase-common with the same name ?

stoty replied:
The same applies to the other "Copy" classes.

NihalJain replied:
Hey @stoty Thanks for having a look at the PR. 

> Do we need two copies of this class ? Can't we just move this from hbase-compression to hbase-common with the same name ?

These files are not supposed to be bundled and are here only for draft change. please see previous comment where I have captured a summary of changes: https://github.com/apache/hbase/pull/6249#issuecomment-2353798881

See
> Created a copy just to bring out the all the usages of the class to help decide (based on references) where we move this in final patch:
RandomDistribution
KeyProviderForTesting
LoadTestKVGenerator

Also see

> Deal with copied classes which have cyclic dependencies and remove all copies and its references
Should we copy these to main of hbase-server or hbase-common?

I have created copy to show case how these files are being used across the project. Just want to discuss if it is fine to move these files to proper module, outside hbase-diagnostics to resolve cyclic dependncies of usage or is it fine to copy them retaining the test copies where ever they are, if possible and avoid copying.

NihalJain replied:
> Can't we just move this from hbase-compression to hbase-common with the same name ?

I am +1 on copying to hbase-common. Will wait some time before updating draft with this change, lets also see what other think about this.

stoty replied:
Can you add a separete patch on top of this one which moves the classes to hbase-common ?
You can open a separate PR for that one, so this one still shows what you wanted.

stoty replied:
I'd like to see how big the patch is without the copies.

NihalJain replied:
> Can you add a separete patch on top of this one which moves the classes to hbase-common ?

Sure @stoty let me post a new PR with suggested change.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   2m 28s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 36s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m  1s |  |  master passed  |
| +1 :green_heart: |  compile  |   2m  4s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   3m 56s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 31s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 52s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   2m  5s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   2m  5s |  |  the patch passed  |
| -0 :warning: |  javadoc  |   0m 12s | [/results-javadoc-javadoc-hbase-diagnostics.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-jdk17-hadoop3-check/output/results-javadoc-javadoc-hbase-diagnostics.txt) |  hbase-diagnostics generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| -0 :warning: |  javadoc  |   2m  4s | [/results-javadoc-javadoc-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-jdk17-hadoop3-check/output/results-javadoc-javadoc-root.txt) |  root generated 2 new + 91 unchanged - 0 fixed = 93 total (was 91)  |
| +1 :green_heart: |  shadedjars  |   6m  7s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 244m  9s | [/patch-unit-root.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-root.txt) |  root in the patch failed.  |
|  |   | 283m 48s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6249 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 75c2ec15443b 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 167c153ab498904c9361cc7d3b4a408ce45267b7 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/testReport/ |
| Max. process+thread count | 5110 (vs. ulimit of 30000) |
| modules | C: hbase-common hbase-client hbase-balancer hbase-asyncfs hbase-server hbase-mapreduce hbase-diagnostics hbase-it hbase-compression/hbase-compression-zstd hbase-assembly . U: . |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6249/3/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

stoty commented on hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestRegionObserverInterface.java:
```diff
@@ -510,29 +509,30 @@ public void testHBase3583() throws IOException {
-    table.close();
-
-  }
+  // TODO: Below test seems to be using FilterAllFilter
```
Need to resolve this before commit

stoty reviewed (commented):
Not having to deal with Copy classes would also make the patch much smaller, half the files only differ due to those renames.

stoty commented:
Sorry I did not read your comments @NihalJain .

NihalJain commented:
> Sorry I did not read your comments @NihalJain .

Hey @stoty no problem at all. I have raised a new PR with suggested changes where I will continue rest of the work, will close this one soon
